<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Подготовка-данных" data-toc-modified-id="Подготовка-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Подготовка данных</a></span><ul class="toc-item"><li><span><a href="#Обзор-данных" data-toc-modified-id="Обзор-данных-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Обзор данных</a></span></li><li><span><a href="#Устранение-пропусков" data-toc-modified-id="Устранение-пропусков-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Устранение пропусков</a></span></li><li><span><a href="#Подготовка-обучающей-и-тестовой-выборок." data-toc-modified-id="Подготовка-обучающей-и-тестовой-выборок.-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Подготовка обучающей и тестовой выборок.</a></span></li><li><span><a href="#Преобразование-категориальных-признаков" data-toc-modified-id="Преобразование-категориальных-признаков-1.4"><span class="toc-item-num">1.4&nbsp;&nbsp;</span>Преобразование категориальных признаков</a></span></li><li><span><a href="#Масштабирование-признаков" data-toc-modified-id="Масштабирование-признаков-1.5"><span class="toc-item-num">1.5&nbsp;&nbsp;</span>Масштабирование признаков</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-1.6"><span class="toc-item-num">1.6&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Исследование-задачи" data-toc-modified-id="Исследование-задачи-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Исследование задачи</a></span><ul class="toc-item"><li><span><a href="#Баланс-классов" data-toc-modified-id="Баланс-классов-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Баланс классов</a></span></li><li><span><a href="#Базовая-модель" data-toc-modified-id="Базовая-модель-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Базовая модель</a></span></li><li><span><a href="#Обучение-моделей-без-учета-дисбаланса" data-toc-modified-id="Обучение-моделей-без-учета-дисбаланса-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Обучение моделей без учета дисбаланса</a></span><ul class="toc-item"><li><span><a href="#Дерево-решений" data-toc-modified-id="Дерево-решений-2.3.1"><span class="toc-item-num">2.3.1&nbsp;&nbsp;</span>Дерево решений</a></span></li><li><span><a href="#Случайный-лес" data-toc-modified-id="Случайный-лес-2.3.2"><span class="toc-item-num">2.3.2&nbsp;&nbsp;</span>Случайный лес</a></span></li><li><span><a href="#Логистическая-регрессия" data-toc-modified-id="Логистическая-регрессия-2.3.3"><span class="toc-item-num">2.3.3&nbsp;&nbsp;</span>Логистическая регрессия</a></span></li></ul></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Борьба-с-дисбалансом" data-toc-modified-id="Борьба-с-дисбалансом-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Борьба с дисбалансом</a></span><ul class="toc-item"><li><span><a href="#Взвешивание-классов" data-toc-modified-id="Взвешивание-классов-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Взвешивание классов</a></span><ul class="toc-item"><li><span><a href="#Дерево-решений" data-toc-modified-id="Дерево-решений-3.1.1"><span class="toc-item-num">3.1.1&nbsp;&nbsp;</span>Дерево решений</a></span></li><li><span><a href="#Случайный-лес" data-toc-modified-id="Случайный-лес-3.1.2"><span class="toc-item-num">3.1.2&nbsp;&nbsp;</span>Случайный лес</a></span></li><li><span><a href="#Логистическая-регрессия" data-toc-modified-id="Логистическая-регрессия-3.1.3"><span class="toc-item-num">3.1.3&nbsp;&nbsp;</span>Логистическая регрессия</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-3.1.4"><span class="toc-item-num">3.1.4&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Увеличение-выборки" data-toc-modified-id="Увеличение-выборки-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Увеличение выборки</a></span><ul class="toc-item"><li><span><a href="#Дерево-решений" data-toc-modified-id="Дерево-решений-3.2.1"><span class="toc-item-num">3.2.1&nbsp;&nbsp;</span>Дерево решений</a></span></li><li><span><a href="#Случайный-лес" data-toc-modified-id="Случайный-лес-3.2.2"><span class="toc-item-num">3.2.2&nbsp;&nbsp;</span>Случайный лес</a></span></li><li><span><a href="#Логистическая-регрессия" data-toc-modified-id="Логистическая-регрессия-3.2.3"><span class="toc-item-num">3.2.3&nbsp;&nbsp;</span>Логистическая регрессия</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-3.2.4"><span class="toc-item-num">3.2.4&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Продвинутое-увеличение-выборки" data-toc-modified-id="Продвинутое-увеличение-выборки-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Продвинутое увеличение выборки</a></span><ul class="toc-item"><li><span><a href="#Дерево-решений" data-toc-modified-id="Дерево-решений-3.3.1"><span class="toc-item-num">3.3.1&nbsp;&nbsp;</span>Дерево решений</a></span></li><li><span><a href="#Случайный-лес" data-toc-modified-id="Случайный-лес-3.3.2"><span class="toc-item-num">3.3.2&nbsp;&nbsp;</span>Случайный лес</a></span></li><li><span><a href="#Логистическая-регрессия" data-toc-modified-id="Логистическая-регрессия-3.3.3"><span class="toc-item-num">3.3.3&nbsp;&nbsp;</span>Логистическая регрессия</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-3.3.4"><span class="toc-item-num">3.3.4&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Уменьшение-выборки" data-toc-modified-id="Уменьшение-выборки-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Уменьшение выборки</a></span><ul class="toc-item"><li><span><a href="#Дерево-решений" data-toc-modified-id="Дерево-решений-3.4.1"><span class="toc-item-num">3.4.1&nbsp;&nbsp;</span>Дерево решений</a></span></li><li><span><a href="#Случайный-лес" data-toc-modified-id="Случайный-лес-3.4.2"><span class="toc-item-num">3.4.2&nbsp;&nbsp;</span>Случайный лес</a></span></li><li><span><a href="#Логистическая-регрессия" data-toc-modified-id="Логистическая-регрессия-3.4.3"><span class="toc-item-num">3.4.3&nbsp;&nbsp;</span>Логистическая регрессия</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-3.4.4"><span class="toc-item-num">3.4.4&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Увеличение-выборки-для-случайного-леса-без-кросс-валидации" data-toc-modified-id="Увеличение-выборки-для-случайного-леса-без-кросс-валидации-3.5"><span class="toc-item-num">3.5&nbsp;&nbsp;</span>Увеличение выборки для случайного леса без кросс-валидации</a></span></li><li><span><a href="#Выводы-по-борьбе-с-дисбалансом" data-toc-modified-id="Выводы-по-борьбе-с-дисбалансом-3.6"><span class="toc-item-num">3.6&nbsp;&nbsp;</span>Выводы по борьбе с дисбалансом</a></span></li></ul></li><li><span><a href="#Тестирование-модели" data-toc-modified-id="Тестирование-модели-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Тестирование модели</a></span><ul class="toc-item"><li><span><a href="#Дерево-решений" data-toc-modified-id="Дерево-решений-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Дерево решений</a></span></li><li><span><a href="#Случайный-лес" data-toc-modified-id="Случайный-лес-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Случайный лес</a></span></li><li><span><a href="#Логистическая-регрессия" data-toc-modified-id="Логистическая-регрессия-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Логистическая регрессия</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-4.4"><span class="toc-item-num">4.4&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Выводы" data-toc-modified-id="Выводы-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Выводы</a></span></li><li><span><a href="#Чек-лист-готовности-проекта" data-toc-modified-id="Чек-лист-готовности-проекта-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Чек-лист готовности проекта</a></span></li></ul></div>

# Отток клиентов

Из «Бета-Банка» стали уходить клиенты. Каждый месяц. Немного, но заметно. Банковские маркетологи посчитали: сохранять текущих клиентов дешевле, чем привлекать новых.

Нужно спрогнозировать, уйдёт клиент из банка в ближайшее время или нет. Вам предоставлены исторические данные о поведении клиентов и расторжении договоров с банком. 

Постройте модель с предельно большим значением *F1*-меры. Чтобы сдать проект успешно, нужно довести метрику до 0.59. Проверьте *F1*-меру на тестовой выборке самостоятельно.

Дополнительно измеряйте *AUC-ROC*, сравнивайте её значение с *F1*-мерой.

Источник данных: [https://www.kaggle.com/barelydedicated/bank-customer-churn-modeling](https://www.kaggle.com/barelydedicated/bank-customer-churn-modeling)

## Подготовка данных

In [44]:

!pip install optuna==3.0.4 imbalanced-learn==0.10.0 scikit-learn==1.1.3

In [45]:
import optuna as optuna
import pandas as pd
from imblearn.over_sampling import SMOTE
from imblearn.pipeline import make_pipeline
from sklearn.dummy import DummyClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import f1_score
from sklearn.metrics import recall_score
from sklearn.metrics import roc_auc_score
from sklearn.model_selection import train_test_split, GridSearchCV, RandomizedSearchCV
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.tree import DecisionTreeClassifier
from sklearn.utils import shuffle

In [46]:
RANDOM_STATE = 12345

### Обзор данных
Произведем первичный обзор данных.

In [47]:
try:
    data = pd.read_csv('/datasets/Churn.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/Churn.csv')
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RowNumber        10000 non-null  int64  
 1   CustomerId       10000 non-null  int64  
 2   Surname          10000 non-null  object 
 3   CreditScore      10000 non-null  int64  
 4   Geography        10000 non-null  object 
 5   Gender           10000 non-null  object 
 6   Age              10000 non-null  int64  
 7   Tenure           9091 non-null   float64
 8   Balance          10000 non-null  float64
 9   NumOfProducts    10000 non-null  int64  
 10  HasCrCard        10000 non-null  int64  
 11  IsActiveMember   10000 non-null  int64  
 12  EstimatedSalary  10000 non-null  float64
 13  Exited           10000 non-null  int64  
dtypes: float64(3), int64(8), object(3)
memory usage: 1.1+ MB


In [48]:
data.head()

RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0     2.0       0.00              1          1               1   
1     1.0   83807.86              1          0               1   
2     8.0  159660.80              3          1               0   
3     1.0       0.00              2          0               0   
4     2.0  125510.82              1          1               1   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         79084.10       0

### Устранение пропусков
В столбце Tenure есть пропуски, посмотрим отдельно на объекты с пропусками.

In [49]:
data[data['Tenure'].isna()]

RowNumber  CustomerId    Surname  CreditScore Geography  Gender  Age  \
30           31    15589475    Azikiwe          591     Spain  Female   39   
48           49    15766205        Yin          550   Germany    Male   38   
51           52    15768193  Trevisani          585   Germany    Male   36   
53           54    15702298   Parkhill          655   Germany    Male   41   
60           61    15651280     Hunter          742   Germany    Male   35   
...         ...         ...        ...          ...       ...     ...  ...   
9944       9945    15703923    Cameron          744   Germany    Male   41   
9956       9957    15707861      Nucci          520    France  Female   46   
9964       9965    15642785    Douglas          479    France    Male   34   
9985       9986    15586914     Nepean          659    France    Male   36   
9999      10000    15628319     Walker          792    France  Female   28   

      Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
30       NaN       0.00              3          1               0   
48       NaN  103391.38              1          0               1   
51       NaN  146050.97              2          0               0   
53       NaN  125561.97              1          0               0   
60       NaN  136857.00              1          0               0   
...      ...        ...            ...        ...             ...   
9944     NaN  190409.34              2          1               1   
9956     NaN   85216.61              1          1               0   
9964     NaN  117593.48              2          0               0   
9985     NaN  123841.49              2          1               0   
9999     NaN  130142.79              1          1               0   

      EstimatedSalary  Exited  
30          140469.38       1  
48           90878.13       0  
51           86424.57       0  
53          164040.94       1  
60           84509.57       0  
...               ...     ...  
9944        138361.48       0  
9956        117369.52       1  
9964        113308.29       0  
9985         96833.00       0  
9999         38190.78       0  

[909 rows x 14 columns]

In [50]:
data['Tenure'].value_counts()

1.0     952
2.0     950
8.0     933
3.0     928
5.0     927
7.0     925
4.0     885
9.0     882
6.0     881
10.0    446
0.0     382
Name: Tenure, dtype: int64

Какой-то закономерности в пропусках заметить не удалось, заменим пропуски на медианное значение.

In [51]:
data['Tenure'] = data['Tenure'].fillna(data['Tenure'].median())
data['Tenure'].value_counts()

5.0     1836
1.0      952
2.0      950
8.0      933
3.0      928
7.0      925
4.0      885
9.0      882
6.0      881
10.0     446
0.0      382
Name: Tenure, dtype: int64

Пропуски устранены.

### Подготовка обучающей и тестовой выборок.
В данных есть столбцы, которые никак не влияют на целевой признак, а являются техническими - это индекс строки и идентификатор клиента. Также сомнительно, что на факт ухода клиента влияет фамилия. Подготовим список столбцов, которые необходимо отбросить, чтобы получить столбцы со значащими признаками.

In [52]:
drop_columns = ['RowNumber', 'CustomerId', 'Surname']
data = data.drop(columns=drop_columns)
data.head()

CreditScore Geography  Gender  Age  Tenure    Balance  NumOfProducts  \
0          619    France  Female   42     2.0       0.00              1   
1          608     Spain  Female   41     1.0   83807.86              1   
2          502    France  Female   42     8.0  159660.80              3   
3          699    France  Female   39     1.0       0.00              2   
4          850     Spain  Female   43     2.0  125510.82              1   

   HasCrCard  IsActiveMember  EstimatedSalary  Exited  
0          1               1        101348.88       1  
1          0               1        112542.58       0  
2          1               0        113931.57       1  
3          0               0         93826.63       0  
4          1               1         79084.10       0

Разобъем данные на обучающие и тестовые.

In [53]:
features_train, features_test, target_train, target_test = train_test_split(
    data.drop('Exited', axis=1), data['Exited'], stratify=data['Exited'], test_size=0.25, random_state=RANDOM_STATE
)
print(features_train.shape)
print(features_test.shape)
print(target_train.shape)
print(target_test.shape)

(7500, 10)
(2500, 10)
(7500,)
(2500,)


### Преобразование категориальных признаков
Для страны проживания и пола применим one-hot кодирование для перевода строк в числа.

In [54]:
categorial_columns = ['Geography', 'Gender']

# drop='first' удаляет первый признак из закодированных:
# таким образом обходим dummy-ловушку
# задаём handle_unknown='ignore':
# игнорируется ранее невстречающиеся значения признака (при transform)
encoder_ohe = OneHotEncoder(drop='first', handle_unknown='ignore', sparse=False)

# обучаем энкодер на заданных категориальных признаках тренировочной выборки
encoder_ohe.fit(features_train[categorial_columns])

# добавляем закодированные признаки в features_train
# encoder_ohe.get_feature_names_out() позволяет получить названия колонок
features_train[
    encoder_ohe.get_feature_names_out()
] = encoder_ohe.transform(features_train[categorial_columns])

# удаляем незакодированные категориальные признаки (изначальные колонки)
features_train = features_train.drop(categorial_columns, axis=1)

features_train.head()

CreditScore  Age  Tenure    Balance  NumOfProducts  HasCrCard  \
3349          665   45     5.0  155447.65              2          1   
7094          640   67     3.0       0.00              1          0   
9204          646   30     5.0       0.00              2          1   
6620          732   36     5.0       0.00              2          1   
4737          791   36     6.0  111168.97              1          1   

      IsActiveMember  EstimatedSalary  Geography_Germany  Geography_Spain  \
3349               0         51871.95                1.0              0.0   
7094               1         42964.63                0.0              0.0   
9204               0         13935.32                0.0              0.0   
6620               0        161428.25                0.0              0.0   
4737               1        189969.91                0.0              0.0   

      Gender_Male  
3349          0.0  
7094          0.0  
9204          0.0  
6620          1.0  
4737          1.0

In [55]:
features_test[
    encoder_ohe.get_feature_names_out()
] = encoder_ohe.transform(features_test[categorial_columns])

features_test = features_test.drop(categorial_columns, axis=1)

features_test.head()

CreditScore  Age  Tenure    Balance  NumOfProducts  HasCrCard  \
7888          594   31     0.0   79340.95              1          1   
1567          713   53     6.0  115029.40              1          0   
8704          784   42     2.0  109052.04              2          1   
7244          445   33     7.0       0.00              2          1   
6466          678   32     4.0  139626.01              1          1   

      IsActiveMember  EstimatedSalary  Geography_Germany  Geography_Spain  \
7888               0         78255.86                0.0              0.0   
1567               0        191521.32                0.0              0.0   
8704               0          6409.55                0.0              1.0   
7244               0        122625.68                0.0              0.0   
6466               1        118235.52                1.0              0.0   

      Gender_Male  
7888          0.0  
1567          1.0  
8704          0.0  
7244          0.0  
6466          1.0

Данные закодированы.

### Масштабирование признаков
Произведем стандартизацию числовых признаков.

In [56]:
scaler = StandardScaler()
numeric = ['CreditScore', 'Age', 'Tenure', 'Balance', 'NumOfProducts', 'EstimatedSalary']
scaler.fit(features_train[numeric])
features_train[numeric] = scaler.transform(features_train[numeric])
features_test[numeric] = scaler.transform(features_test[numeric])
features_train.head()

CreditScore       Age    Tenure   Balance  NumOfProducts  HasCrCard  \
3349     0.153616  0.583435 -0.004794  1.264113       0.807524          1   
7094    -0.104283  2.694224 -0.731152 -1.222762      -0.915010          0   
9204    -0.042388 -0.855740 -0.004794 -1.222762       0.807524          1   
6620     0.844786 -0.280070 -0.004794 -1.222762       0.807524          1   
4737     1.453428 -0.280070  0.358385  0.555736      -0.915010          1   

      IsActiveMember  EstimatedSalary  Geography_Germany  Geography_Spain  \
3349               0        -0.848725                1.0              0.0   
7094               1        -1.003683                0.0              0.0   
9204               0        -1.508697                0.0              0.0   
6620               0         1.057193                0.0              0.0   
4737               1         1.553724                0.0              0.0   

      Gender_Male  
3349          0.0  
7094          0.0  
9204          0.0  
6620          1.0  
4737          1.0

Признаки стандартизированы.

### Вывод
Произведен обзор данных и подготовка данных - устранены пропуски, устранены не влияющие на целевую переменную признаки, числовые признаки стандартизированы, преобразованы категориальные признаки, подготовлены обучающая и тестовая выборки.

## Исследование задачи

### Баланс классов
Посмотрим на количество значений целевой переменной.

In [57]:
data['Exited'].value_counts()

0    7963
1    2037
Name: Exited, dtype: int64

Наблюдается существенный дисбаланс - оставшихся клиентов почти в 4 раза больше чем оставшихся.

### Базовая модель

Так как стоит задача максимизировать метрику, показывающую успешность предсказывания единицы, в качестве базовой модели возьмем модель, всегда предсказывающую 1.

In [58]:
dummy_model = DummyClassifier(strategy='constant', constant=1)
dummy_model.fit(features_train, target_train)
print(f'f1 = {f1_score(dummy_model.predict(features_test), target_test)}')

f1 = 0.3383183781987371


Метрика f1 базовой модели равна 0.338, в дальнейшем будет сравнивать метрики обученных моделей с ней

### Обучение моделей без учета дисбаланса
Найдем оптимальные гиперпараметры для дерева решений, случайного леса и логистической регрессии с целью максимизации метрики f1, также будем смотреть на метрику AUC-ROC, для ее вычисления напишем функцию.

In [59]:
def roc_auc(model, features, target):
    probabilities = model.predict_proba(features)
    probabilities_ones = probabilities[:, 1]
    return roc_auc_score(target, probabilities_ones)

#### Дерево решений

In [60]:
TREE_N_ITER = 20

In [61]:
decision_tree_model = DecisionTreeClassifier(random_state=RANDOM_STATE)
decision_tree_params = {
    'max_depth': range(1, 21),
    'min_samples_split': range(2, 11),
    'min_samples_leaf': range(1, 11)
}

grid = RandomizedSearchCV(decision_tree_model, decision_tree_params, scoring='f1', n_jobs=-1, verbose=1, n_iter=TREE_N_ITER, random_state=RANDOM_STATE)
grid.fit(features_train, target_train)
decision_tree_model_disbalance = grid.best_estimator_
print(f'f1 = {grid.best_score_}')
print(f'AUC-ROC = {roc_auc(decision_tree_model_disbalance, features_train, target_train)}')
print(f'Оптимальные гиперпараметры: {grid.best_params_}')

Fitting 5 folds for each of 20 candidates, totalling 100 fits
f1 = 0.5583407125650095
AUC-ROC = 0.9483870847550349
Оптимальные гиперпараметры: {'min_samples_split': 7, 'min_samples_leaf': 9, 'max_depth': 19}


#### Случайный лес

Для ускорения подбора гиперпараметров воспользуемся библиотекой optuna

In [62]:
OPTUNA_N_TRIALS = 20
OPTUNA_TIMEOUT = 30

In [63]:
random_forest_model = RandomForestClassifier(random_state=RANDOM_STATE)
random_forest_params = {
    'n_estimators': optuna.distributions.IntDistribution(10, 500, step=10),
    'max_depth': optuna.distributions.IntDistribution(1, 100),
    'min_samples_split': optuna.distributions.IntDistribution(2, 10),
    'min_samples_leaf': optuna.distributions.IntDistribution(1, 10)
}

optuna_search = optuna.integration.OptunaSearchCV(random_forest_model, random_forest_params, scoring='f1', n_trials=OPTUNA_N_TRIALS, timeout=OPTUNA_TIMEOUT, n_jobs=-1)
optuna_search.fit(features_train, target_train)
random_forest_model_disbalance = optuna_search.best_estimator_
print(f'f1 = {optuna_search.best_score_}')
print(f'AUC-ROC = {roc_auc(random_forest_model_disbalance, features_train, target_train)}')
print(f'Оптимальные гиперпараметры: {optuna_search.best_params_}')

/var/folders/m_/d5s2gnsx68l769b4nc3pxzrdn_qnjl/T/ipykernel_69864/30791300.py:9: ExperimentalWarning: OptunaSearchCV is experimental (supported from v0.17.0). The interface can change in the future.
  optuna_search = optuna.integration.OptunaSearchCV(random_forest_model, random_forest_params, scoring='f1', n_trials=OPTUNA_N_TRIALS, timeout=OPTUNA_TIMEOUT, n_jobs=-1)
[I 2023-03-07 20:01:16,116] A new study created in memory with name: no-name-c0dde356-c9e9-409b-b724-29c4d9a9b039
[I 2023-03-07 20:01:20,329] Trial 7 finished with value: 0.5777796095607051 and parameters: {'n_estimators': 20, 'max_depth': 32, 'min_samples_split': 5, 'min_samples_leaf': 2}. Best is trial 7 with value: 0.5777796095607051.
[I 2023-03-07 20:01:22,102] Trial 12 finished with value: 0.5648784393041888 and parameters: {'n_estimators': 10, 'max_depth': 56, 'min_samples_split': 6, 'min_samples_leaf': 5}. Best is trial 7 with value: 0.5777796095607051.
[I 2023-03-07 20:01:33,618] Trial 3 finished with value: 0.570767

f1 = 0.5790718416031755
AUC-ROC = 0.9807610033559753
Оптимальные гиперпараметры: {'n_estimators': 360, 'max_depth': 13, 'min_samples_split': 9, 'min_samples_leaf': 1}


#### Логистическая регрессия

In [64]:
logistic_regression_model = LogisticRegression(random_state=RANDOM_STATE)
logistic_regression_params = {
    'solver': ['liblinear'],
    'max_iter': range(30, 201)
}

grid = GridSearchCV(logistic_regression_model, logistic_regression_params, scoring='f1', n_jobs=-1, verbose=1)
grid.fit(features_train, target_train)
logistic_regression_model_disbalance = grid.best_estimator_
print(f'f1 = {grid.best_score_}')
print(f'AUC-ROC = {roc_auc(logistic_regression_model_disbalance, features_train, target_train)}')
print(f'Оптимальные гиперпараметры: {grid.best_params_}')

Fitting 5 folds for each of 171 candidates, totalling 855 fits
f1 = 0.31026307581536294
AUC-ROC = 0.7637226340724428
Оптимальные гиперпараметры: {'max_iter': 30, 'solver': 'liblinear'}


### Вывод
На несбалансированных классах удалось достигнуть значения метрики f1 равного 0.586 при помощи модели случайного леса. Логистическая регрессия показала результат хуже, чем базовая модель, всегда предсказывающая 1. Остальные модели показали почти двухкратное увеличение метрики по сравнению с базовой.

|           | Дерево решений                                                                   | Случайный лес                                                                                          | Логистическая регрессия                                 |
|-----------|----------------------------------------------------------------------------------|--------------------------------------------------------------------------------------------------------|---------------------------------------------------------|
| Дисбаланс | 0.566/0.877<br />{'max_depth': 7, 'min_samples_leaf': 2, 'min_samples_split': 8} | 0.586/0.998<br />{'n_estimators': 190, 'max_depth': 20, 'min_samples_split': 6, 'min_samples_leaf': 1} | 0.31/0.764<br />{'max_iter': 30, 'solver': 'liblinear'} |

## Борьба с дисбалансом
Воспользуемся несколькими техниками борьбы с дисбалансом - взвешивание классов, увеличение выборки, уменьшение выборки, дополнительно попробуем продвинутую технику увеличения выборки при помощи библиотеки imblearn.

### Взвешивание классов

#### Дерево решений

In [65]:
decision_tree_model = DecisionTreeClassifier(random_state=RANDOM_STATE, class_weight='balanced')
grid = RandomizedSearchCV(decision_tree_model, decision_tree_params, scoring='f1', n_jobs=-1, verbose=1, n_iter=TREE_N_ITER, random_state=RANDOM_STATE)
grid.fit(features_train, target_train)
decision_tree_model_balance = grid.best_estimator_
print(f'f1 = {grid.best_score_}')
print(f'AUC-ROC = {roc_auc(decision_tree_model_balance, features_train, target_train)}')
print(f'Оптимальные гиперпараметры: {grid.best_params_}')

Fitting 5 folds for each of 20 candidates, totalling 100 fits
f1 = 0.5711991610466954
AUC-ROC = 0.8801815211826218
Оптимальные гиперпараметры: {'min_samples_split': 3, 'min_samples_leaf': 8, 'max_depth': 7}


#### Случайный лес

In [66]:
random_forest_model = RandomForestClassifier(random_state=RANDOM_STATE, class_weight='balanced')
optuna_search = optuna.integration.OptunaSearchCV(random_forest_model, random_forest_params, scoring='f1', n_trials=OPTUNA_N_TRIALS, timeout=OPTUNA_TIMEOUT, n_jobs=-1)
optuna_search.fit(features_train, target_train)
random_forest_model_balance = optuna_search.best_estimator_
print(f'f1 = {optuna_search.best_score_}')
print(f'AUC-ROC = {roc_auc(random_forest_model_balance, features_train, target_train)}')
print(f'Оптимальные гиперпараметры: {optuna_search.best_params_}')

/var/folders/m_/d5s2gnsx68l769b4nc3pxzrdn_qnjl/T/ipykernel_69864/4293178993.py:2: ExperimentalWarning: OptunaSearchCV is experimental (supported from v0.17.0). The interface can change in the future.
  optuna_search = optuna.integration.OptunaSearchCV(random_forest_model, random_forest_params, scoring='f1', n_trials=OPTUNA_N_TRIALS, timeout=OPTUNA_TIMEOUT, n_jobs=-1)
[I 2023-03-07 20:02:10,446] A new study created in memory with name: no-name-46f33d24-819a-4158-b962-f87971bfb102
[I 2023-03-07 20:02:23,318] Trial 4 finished with value: 0.6152186391738691 and parameters: {'n_estimators': 100, 'max_depth': 38, 'min_samples_split': 7, 'min_samples_leaf': 10}. Best is trial 4 with value: 0.6152186391738691.
[I 2023-03-07 20:02:41,726] Trial 1 finished with value: 0.620051751630364 and parameters: {'n_estimators': 210, 'max_depth': 53, 'min_samples_split': 9, 'min_samples_leaf': 3}. Best is trial 1 with value: 0.620051751630364.
[I 2023-03-07 20:02:42,431] Trial 5 finished with value: 0.6140

f1 = 0.6217657374649076
AUC-ROC = 0.9857849940209634
Оптимальные гиперпараметры: {'n_estimators': 320, 'max_depth': 76, 'min_samples_split': 3, 'min_samples_leaf': 4}


#### Логистическая регрессия

In [67]:
logistic_regression_model = LogisticRegression(random_state=RANDOM_STATE, class_weight='balanced')
grid = GridSearchCV(logistic_regression_model, logistic_regression_params, scoring='f1', n_jobs=-1, verbose=1)
grid.fit(features_train, target_train)
logistic_regression_model_balance = grid.best_estimator_
print(f'f1 = {grid.best_score_}')
print(f'AUC-ROC = {roc_auc(logistic_regression_model_balance, features_train, target_train)}')
print(f'Оптимальные гиперпараметры: {grid.best_params_}')

Fitting 5 folds for each of 171 candidates, totalling 855 fits
f1 = 0.4838018984587868
AUC-ROC = 0.7668917645346696
Оптимальные гиперпараметры: {'max_iter': 30, 'solver': 'liblinear'}


#### Вывод
С помощью взвешивания классов удалось достигнуть значения метрики f1 равного 0.624 при помощи модели случайного леса. Для остальных моделей метрика также увеличилась и все модели превзошли константную модель. С ростом f1 также наблюдается рост AUC-ROC.

|             | Дерево решений                                                                   | Случайный лес                                                                                           | Логистическая регрессия                                  |
|-------------|----------------------------------------------------------------------------------|---------------------------------------------------------------------------------------------------------|----------------------------------------------------------|
| Дисбаланс   | 0.566/0.877<br />{'max_depth': 7, 'min_samples_leaf': 2, 'min_samples_split': 8} | 0.586/0.998<br />{'n_estimators': 190, 'max_depth': 20, 'min_samples_split': 6, 'min_samples_leaf': 1}  | 0.31/0.764<br />{'max_iter': 30, 'solver': 'liblinear'}  |
| Взвешивание | 0.574/0.881<br />{'max_depth': 7, 'min_samples_leaf': 6, 'min_samples_split': 2} | 0.624/0.982<br />{'n_estimators': 190, 'max_depth': 16, 'min_samples_split': 10, 'min_samples_leaf': 4} | 0.484/0.767<br />{'max_iter': 30, 'solver': 'liblinear'} |

### Увеличение выборки
Количество единиц в целевой переменной в 4 раза меньше чем нулей, увеличим их количество в 4 раза.

In [68]:
def upsample(features, target, repeat):
    features_zeros = features[target == 0]
    features_ones = features[target == 1]
    target_zeros = target[target == 0]
    target_ones = target[target == 1]

    features_upsampled = pd.concat([features_zeros] + [features_ones] * repeat)
    target_upsampled = pd.concat([target_zeros] + [target_ones] * repeat)

    features_upsampled, target_upsampled = shuffle(features_upsampled, target_upsampled, random_state=RANDOM_STATE)

    return features_upsampled, target_upsampled

In [69]:
train_features_upsampled, train_target_upsampled = upsample(features_train, target_train, 4)

Простое увеличение выборки обычно некорректно работает с кросс-валидацией из-за утечки данных. Проверим как эти подходы отработают на текущих данных.

#### Дерево решений

In [70]:
decision_tree_model = DecisionTreeClassifier(random_state=RANDOM_STATE)
grid = RandomizedSearchCV(decision_tree_model, decision_tree_params, scoring='f1', n_jobs=-1, verbose=1, n_iter=TREE_N_ITER, random_state=RANDOM_STATE)
grid.fit(train_features_upsampled, train_target_upsampled)
decision_tree_model_upsample = grid.best_estimator_
print(f'f1 = {grid.best_score_}')
print(f'AUC-ROC = {roc_auc(decision_tree_model_upsample, features_train, target_train)}')
print(f'Оптимальные гиперпараметры: {grid.best_params_}')

Fitting 5 folds for each of 20 candidates, totalling 100 fits
f1 = 0.9170005864052225
AUC-ROC = 0.9979757191501002
Оптимальные гиперпараметры: {'min_samples_split': 2, 'min_samples_leaf': 1, 'max_depth': 16}


#### Случайный лес

In [71]:
random_forest_model = RandomForestClassifier(random_state=RANDOM_STATE)
optuna_search = optuna.integration.OptunaSearchCV(random_forest_model, random_forest_params, scoring='f1', n_trials=OPTUNA_N_TRIALS, timeout=OPTUNA_TIMEOUT, n_jobs=-1, verbose=1)
optuna_search.fit(train_features_upsampled, train_target_upsampled)
random_forest_model_upsample = optuna_search.best_estimator_
print(f'f1 = {optuna_search.best_score_}')
print(f'AUC-ROC = {roc_auc(random_forest_model_upsample, features_train, target_train)}')
print(f'Оптимальные гиперпараметры: {optuna_search.best_params_}')

/var/folders/m_/d5s2gnsx68l769b4nc3pxzrdn_qnjl/T/ipykernel_69864/1844268609.py:2: ExperimentalWarning: OptunaSearchCV is experimental (supported from v0.17.0). The interface can change in the future.
  optuna_search = optuna.integration.OptunaSearchCV(random_forest_model, random_forest_params, scoring='f1', n_trials=OPTUNA_N_TRIALS, timeout=OPTUNA_TIMEOUT, n_jobs=-1, verbose=1)
[I 2023-03-07 20:03:04,093] A new study created in memory with name: no-name-64f928f3-ad7b-44af-b7be-654b8db2d45d
[I 2023-03-07 20:03:04,094] Searching the best hyperparameters using 12084 samples...
[I 2023-03-07 20:03:15,592] Trial 9 finished with value: 0.8843820045146794 and parameters: {'n_estimators': 70, 'max_depth': 21, 'min_samples_split': 2, 'min_samples_leaf': 6}. Best is trial 9 with value: 0.8843820045146794.
[I 2023-03-07 20:03:27,213] Trial 3 finished with value: 0.9292919667711281 and parameters: {'n_estimators': 130, 'max_depth': 83, 'min_samples_split': 6, 'min_samples_leaf': 3}. Best is trial 

f1 = 0.9436358270697612
AUC-ROC = 1.0
Оптимальные гиперпараметры: {'n_estimators': 170, 'max_depth': 46, 'min_samples_split': 2, 'min_samples_leaf': 2}


#### Логистическая регрессия

In [72]:
logistic_regression_model = LogisticRegression(random_state=RANDOM_STATE)
grid = GridSearchCV(logistic_regression_model, logistic_regression_params, scoring='f1', n_jobs=-1, verbose=1)
grid.fit(train_features_upsampled, train_target_upsampled)
logistic_regression_model_upsample = grid.best_estimator_
print(f'f1 = {grid.best_score_}')
print(f'AUC-ROC = {roc_auc(logistic_regression_model_upsample, features_train, target_train)}')
print(f'Оптимальные гиперпараметры: {grid.best_params_}')

Fitting 5 folds for each of 171 candidates, totalling 855 fits
f1 = 0.6983428102778019
AUC-ROC = 0.7669186132141967
Оптимальные гиперпараметры: {'max_iter': 30, 'solver': 'liblinear'}


#### Вывод
С помощью увеличения выборки удалось достигнуть значения метрики f1 равного 0.958 при помощи модели случайного леса. AUC-ROC вобще достигла максимума - 1. Для остальных моделей метрики также увеличилась. Однако есть подозрение, что дублирование данных привело к переобучению некоторых моделей и они слишком хорошо стали предсказывать единицу. Проверим это в последующем на тестовой выборке.

|             | Дерево решений                                                                    | Случайный лес                                                                                           | Логистическая регрессия                                  |
|-------------|-----------------------------------------------------------------------------------|---------------------------------------------------------------------------------------------------------|----------------------------------------------------------|
| Дисбаланс   | 0.566/0.877<br />{'max_depth': 7, 'min_samples_leaf': 2, 'min_samples_split': 8}  | 0.586/0.998<br />{'n_estimators': 190, 'max_depth': 20, 'min_samples_split': 6, 'min_samples_leaf': 1}  | 0.31/0.764<br />{'max_iter': 30, 'solver': 'liblinear'}  |
| Взвешивание | 0.574/0.881<br />{'max_depth': 7, 'min_samples_leaf': 6, 'min_samples_split': 2}  | 0.624/0.982<br />{'n_estimators': 190, 'max_depth': 16, 'min_samples_split': 10, 'min_samples_leaf': 4} | 0.484/0.767<br />{'max_iter': 30, 'solver': 'liblinear'} |
| Увеличение  | 0.931/0.999<br />{'max_depth': 20, 'min_samples_leaf': 1, 'min_samples_split': 2} | 0.958/1<br />{'n_estimators': 140, 'max_depth': 18, 'min_samples_split': 2, 'min_samples_leaf': 1}      | 0.698/0.767<br />{'max_iter': 30, 'solver': 'liblinear'} |

### Продвинутое увеличение выборки
Воспользуемся функционалом библиотеки imblearn для увеличения выборки.

#### Дерево решений

In [73]:
imba_pipeline = make_pipeline(
    SMOTE(random_state=RANDOM_STATE),
    DecisionTreeClassifier(random_state=RANDOM_STATE)
)

decision_tree_pipline_params = {'decisiontreeclassifier__' + key: value for key, value in decision_tree_params.items()}
grid = RandomizedSearchCV(imba_pipeline, decision_tree_pipline_params, scoring='f1', n_jobs=-1, verbose=1, n_iter=TREE_N_ITER, random_state=RANDOM_STATE)
grid.fit(features_train, target_train)
decision_tree_model_imba = grid.best_estimator_
print(f'f1 = {grid.best_score_}')
print(f'AUC-ROC = {roc_auc(decision_tree_model_imba, features_train, target_train)}')
print(f'Оптимальные гиперпараметры: {grid.best_params_}')

Fitting 5 folds for each of 20 candidates, totalling 100 fits
f1 = 0.5669045730669006
AUC-ROC = 0.8569419069093815
Оптимальные гиперпараметры: {'decisiontreeclassifier__min_samples_split': 3, 'decisiontreeclassifier__min_samples_leaf': 8, 'decisiontreeclassifier__max_depth': 7}


#### Случайный лес

In [74]:
imba_pipeline = make_pipeline(
    SMOTE(random_state=RANDOM_STATE), RandomForestClassifier(random_state=RANDOM_STATE)
)

random_forest_pipeline_params = {'randomforestclassifier__' + key: value for key, value in random_forest_params.items()}
optuna_search = optuna.integration.OptunaSearchCV(
    imba_pipeline, random_forest_pipeline_params, scoring='f1', n_trials=OPTUNA_N_TRIALS, timeout=OPTUNA_TIMEOUT, n_jobs=-1, verbose=1
)
optuna_search.fit(features_train, target_train)
random_forest_model_imba = optuna_search.best_estimator_
print(f'f1 = {optuna_search.best_score_}')
print(f'AUC-ROC = {roc_auc(random_forest_model_imba, features_train, target_train)}')
print(f'Оптимальные гиперпараметры: {optuna_search.best_params_}')

/var/folders/m_/d5s2gnsx68l769b4nc3pxzrdn_qnjl/T/ipykernel_69864/3078620761.py:6: ExperimentalWarning: OptunaSearchCV is experimental (supported from v0.17.0). The interface can change in the future.
  optuna_search = optuna.integration.OptunaSearchCV(
[I 2023-03-07 20:04:19,333] A new study created in memory with name: no-name-78576e84-42e5-4457-8134-1fdb4e602373
[I 2023-03-07 20:04:19,333] Searching the best hyperparameters using 7500 samples...
[I 2023-03-07 20:04:41,142] Trial 0 finished with value: 0.6131792167462967 and parameters: {'randomforestclassifier__n_estimators': 120, 'randomforestclassifier__max_depth': 18, 'randomforestclassifier__min_samples_split': 4, 'randomforestclassifier__min_samples_leaf': 6}. Best is trial 0 with value: 0.6131792167462967.
[I 2023-03-07 20:04:42,864] Trial 10 finished with value: 0.6102921493719007 and parameters: {'randomforestclassifier__n_estimators': 130, 'randomforestclassifier__max_depth': 52, 'randomforestclassifier__min_samples_split': 

f1 = 0.6172885402814687
AUC-ROC = 0.9785260973548462
Оптимальные гиперпараметры: {'randomforestclassifier__n_estimators': 230, 'randomforestclassifier__max_depth': 79, 'randomforestclassifier__min_samples_split': 10, 'randomforestclassifier__min_samples_leaf': 4}


#### Логистическая регрессия

In [75]:
imba_pipeline = make_pipeline(
    SMOTE(random_state=RANDOM_STATE), LogisticRegression(random_state=RANDOM_STATE)
)

logistic_regression_pipeline_params = {'logisticregression__' + key: value for key, value in logistic_regression_params.items()}
grid = GridSearchCV(imba_pipeline, logistic_regression_pipeline_params, scoring='f1', n_jobs=-1, verbose=1)
grid.fit(features_train, target_train)
logistic_regression_model_imba = grid.best_estimator_
print(f'f1 = {grid.best_score_}')
print(f'AUC-ROC = {roc_auc(logistic_regression_model_imba, features_train, target_train)}')
print(f'Оптимальные гиперпараметры: {grid.best_params_}')

Fitting 5 folds for each of 171 candidates, totalling 855 fits
f1 = 0.4731174848650486
AUC-ROC = 0.7611321200506378
Оптимальные гиперпараметры: {'logisticregression__max_iter': 30, 'logisticregression__solver': 'liblinear'}


#### Вывод
С помощью продвинутого увеличения выборки удалось достигнуть значения метрики f1 равного 0.618 при помощи модели случайного леса. Для остальных моделей метрика также увеличилась. Посмотртим как покажут себя обученные модели на тестовой выборке.

|                | Дерево решений                                                                    | Случайный лес                                                                                           | Логистическая регрессия                                  |
|----------------|-----------------------------------------------------------------------------------|---------------------------------------------------------------------------------------------------------|----------------------------------------------------------|
| Дисбаланс      | 0.566/0.877<br />{'max_depth': 7, 'min_samples_leaf': 2, 'min_samples_split': 8}  | 0.586/0.998<br />{'n_estimators': 190, 'max_depth': 20, 'min_samples_split': 6, 'min_samples_leaf': 1}  | 0.31/0.764<br />{'max_iter': 30, 'solver': 'liblinear'}  |
| Взвешивание    | 0.574/0.881<br />{'max_depth': 7, 'min_samples_leaf': 6, 'min_samples_split': 2}  | 0.624/0.982<br />{'n_estimators': 190, 'max_depth': 16, 'min_samples_split': 10, 'min_samples_leaf': 4} | 0.484/0.767<br />{'max_iter': 30, 'solver': 'liblinear'} |
| Увеличение     | 0.931/0.999<br />{'max_depth': 20, 'min_samples_leaf': 1, 'min_samples_split': 2} | 0.958/1<br />{'n_estimators': 140, 'max_depth': 18, 'min_samples_split': 2, 'min_samples_leaf': 1}      | 0.698/0.767<br />{'max_iter': 30, 'solver': 'liblinear'} |
| imblearn SMOTE | 0.571/0.873<br />{'max_depth': 8, 'min_samples_leaf': 9, 'min_samples_split': 2}  | 0.618/0.973<br />{'n_estimators': 170, 'max_depth': 19, 'min_samples_split': 7, 'min_samples_leaf': 5}  | 0.473/0.761<br />{'max_iter': 30, 'solver': 'liblinear'} |

### Уменьшение выборки
Количество нулей в целевой переменной в 4 раза больше чем единиц, уменьшим их количество в 4 раза.

In [76]:
def downsample(features, target, fraction):
    features_zeros = features[target == 0]
    features_ones = features[target == 1]
    target_zeros = target[target == 0]
    target_ones = target[target == 1]

    features_downsampled = pd.concat([features_zeros.sample(frac=fraction, random_state=RANDOM_STATE)] + [features_ones])
    target_downsampled = pd.concat([target_zeros.sample(frac=fraction, random_state=RANDOM_STATE)] + [target_ones])

    features_downsampled, target_downsampled = shuffle(features_downsampled, target_downsampled, random_state=RANDOM_STATE)

    return features_downsampled, target_downsampled

In [77]:
train_features_downsampled, train_target_downsampled = downsample(features_train, target_train, 0.25)

#### Дерево решений

In [78]:
decision_tree_model = DecisionTreeClassifier(random_state=RANDOM_STATE)
grid = RandomizedSearchCV(decision_tree_model, decision_tree_params, scoring='f1', n_jobs=-1, verbose=1, n_iter=TREE_N_ITER, random_state=RANDOM_STATE)
grid.fit(train_features_downsampled, train_target_downsampled)
decision_tree_model_downsample = grid.best_estimator_
print(f'f1 = {grid.best_score_}')
print(f'AUC-ROC = {roc_auc(decision_tree_model_downsample, features_train, target_train)}')
print(f'Оптимальные гиперпараметры: {grid.best_params_}')

Fitting 5 folds for each of 20 candidates, totalling 100 fits
f1 = 0.7666439277879722
AUC-ROC = 0.8662936307480282
Оптимальные гиперпараметры: {'min_samples_split': 2, 'min_samples_leaf': 4, 'max_depth': 7}


#### Случайный лес

In [79]:
random_forest_model = RandomForestClassifier(random_state=RANDOM_STATE)
optuna_search = optuna.integration.OptunaSearchCV(random_forest_model, random_forest_params, scoring='f1', n_trials=200, timeout=OPTUNA_TIMEOUT, n_jobs=-1, verbose=1)
optuna_search.fit(train_features_downsampled, train_target_downsampled)
random_forest_model_downsample = optuna_search.best_estimator_
print(f'f1 = {optuna_search.best_score_}')
print(f'AUC-ROC = {roc_auc(random_forest_model_downsample, features_train, target_train)}')
print(f'Оптимальные гиперпараметры: {optuna_search.best_params_}')

/var/folders/m_/d5s2gnsx68l769b4nc3pxzrdn_qnjl/T/ipykernel_69864/4198472022.py:2: ExperimentalWarning: OptunaSearchCV is experimental (supported from v0.17.0). The interface can change in the future.
  optuna_search = optuna.integration.OptunaSearchCV(random_forest_model, random_forest_params, scoring='f1', n_trials=200, timeout=OPTUNA_TIMEOUT, n_jobs=-1, verbose=1)
[I 2023-03-07 20:05:44,741] A new study created in memory with name: no-name-22421769-c536-423e-9548-881d9e60d960
[I 2023-03-07 20:05:44,742] Searching the best hyperparameters using 3021 samples...
[I 2023-03-07 20:05:55,513] Trial 0 finished with value: 0.7749091015224391 and parameters: {'n_estimators': 70, 'max_depth': 50, 'min_samples_split': 4, 'min_samples_leaf': 2}. Best is trial 0 with value: 0.7749091015224391.
[I 2023-03-07 20:05:56,950] Trial 3 finished with value: 0.7730289496566625 and parameters: {'n_estimators': 80, 'max_depth': 19, 'min_samples_split': 2, 'min_samples_leaf': 3}. Best is trial 0 with value: 

f1 = 0.7778273998065843
AUC-ROC = 0.9100162670122
Оптимальные гиперпараметры: {'n_estimators': 310, 'max_depth': 41, 'min_samples_split': 4, 'min_samples_leaf': 8}


#### Логистическая регрессия

In [80]:
logistic_regression_model = LogisticRegression(random_state=RANDOM_STATE)
grid = GridSearchCV(logistic_regression_model, logistic_regression_params, scoring='f1', n_jobs=-1, verbose=1)
grid.fit(train_features_downsampled, train_target_downsampled)
logistic_regression_model_downsample = grid.best_estimator_
print(f'f1 = {grid.best_score_}')
print(f'AUC-ROC = {roc_auc(logistic_regression_model_downsample, features_train, target_train)}')
print(f'Оптимальные гиперпараметры: {grid.best_params_}')

Fitting 5 folds for each of 171 candidates, totalling 855 fits
f1 = 0.7037559136230817
AUC-ROC = 0.7672975631480943
Оптимальные гиперпараметры: {'max_iter': 30, 'solver': 'liblinear'}


#### Вывод
С помощью уменьшения выборки удалось достигнуть значения метрики f1 равного 0.78 при помощи модели случайного леса. Для остальных моделей метрика также увеличилась по сравнению с несбалансированными выборками и взвешиванием классов, а для логистической регрессии уменьшение выборки дало самое большое значение метрики. Однако также, как и с увеличением выборки, есть подозрение, что на тестовой выборке результаты будут хуже.

|                | Дерево решений                                                                    | Случайный лес                                                                                           | Логистическая регрессия                                  |
|----------------|-----------------------------------------------------------------------------------|---------------------------------------------------------------------------------------------------------|----------------------------------------------------------|
| Дисбаланс      | 0.566/0.877<br />{'max_depth': 7, 'min_samples_leaf': 2, 'min_samples_split': 8}  | 0.586/0.998<br />{'n_estimators': 190, 'max_depth': 20, 'min_samples_split': 6, 'min_samples_leaf': 1}  | 0.31/0.764<br />{'max_iter': 30, 'solver': 'liblinear'}  |
| Взвешивание    | 0.574/0.881<br />{'max_depth': 7, 'min_samples_leaf': 6, 'min_samples_split': 2}  | 0.624/0.982<br />{'n_estimators': 190, 'max_depth': 16, 'min_samples_split': 10, 'min_samples_leaf': 4} | 0.484/0.767<br />{'max_iter': 30, 'solver': 'liblinear'} |
| Увеличение     | 0.931/0.999<br />{'max_depth': 20, 'min_samples_leaf': 1, 'min_samples_split': 2} | 0.958/1<br />{'n_estimators': 140, 'max_depth': 18, 'min_samples_split': 2, 'min_samples_leaf': 1}      | 0.698/0.767<br />{'max_iter': 30, 'solver': 'liblinear'} |
| imblearn SMOTE | 0.571/0.873<br />{'max_depth': 8, 'min_samples_leaf': 9, 'min_samples_split': 2}  | 0.618/0.973<br />{'n_estimators': 170, 'max_depth': 19, 'min_samples_split': 7, 'min_samples_leaf': 5}  | 0.473/0.761<br />{'max_iter': 30, 'solver': 'liblinear'} |
| Уменьшение     | 0.769/0.866<br />{'max_depth': 7, 'min_samples_leaf': 10, 'min_samples_split': 2} | 0.779/0.920<br />{'n_estimators': 100, 'max_depth': 13, 'min_samples_split': 6, 'min_samples_leaf': 6}  | 0.704/0.767<br />{'max_iter': 30, 'solver': 'liblinear'} |

### Увеличение выборки для случайного леса без кросс-валидации
Так как увеличение выборки может некорректно работать с кросс-валидацие, обучим одну из моделей при помощи обучающей и валидационной выборок и перебором гиперпараметров в цикле. До этого лучше всего себя показывал случайны лес, возьмем его.

In [81]:
features_train_upsample, features_valid_upsample, target_train_upsample, target_valid_upsample = train_test_split(
    features_train, target_train, stratify=target_train, test_size=0.25, random_state=RANDOM_STATE
)
print(features_train_upsample.shape)
print(features_valid_upsample.shape)
print(target_train_upsample.shape)
print(target_valid_upsample.shape)

features_train_upsample, target_train_upsample = upsample(features_train_upsample, target_train_upsample, 4)

print(features_train_upsample.shape)
print(features_valid_upsample.shape)
print(target_train_upsample.shape)
print(target_valid_upsample.shape)

print(target_train_upsample.mean())
print(target_valid_upsample.mean())

(5625, 11)
(1875, 11)
(5625,)
(1875,)
(9063, 11)
(1875, 11)
(9063,)
(1875,)
0.5057927838464085
0.20373333333333332


In [82]:
from joblib import Parallel, delayed

def trial(n_estimators, max_depth, min_samples_split, min_samples_leaf, random_state, X, y, X_valid, y_valid):
    params = {
        'n_estimators': n_estimators,
        'max_depth': max_depth,
        'min_samples_split': min_samples_split,
        'min_samples_leaf': min_samples_leaf,
    }
    print(params)
    random_forest_model = RandomForestClassifier(
        **params,
        random_state=random_state
    )
    random_forest_model.fit(X, y)
    f1 = f1_score(y_valid, random_forest_model.predict(X_valid))
    return f1, params, random_forest_model


scores = Parallel(n_jobs=-1)(delayed(trial)(
    n_estimators, max_depth, min_samples_split, min_samples_leaf, RANDOM_STATE, features_train_upsample,
    target_train_upsample, features_valid_upsample, target_valid_upsample
)
    for n_estimators in range(10, 201, 20)
    for max_depth in range(1, 21, 2)
    for min_samples_split in range(2, 11, 2)
    for min_samples_leaf in range(2, 11, 2)
)

f1_list = [score[0] for score in scores]
params_list = [score[1] for score in scores]
model_list = [score[2] for score in scores]
best_f1 = max(f1_list)
max_index = f1_list.index(best_f1)
best_params = params_list[max_index]
random_forest_model_no_cv = model_list[max_index]

print(f'f1 = {best_f1}')
print(f'AUC-ROC = {roc_auc(random_forest_model_no_cv, features_valid_upsample, target_valid_upsample)}')
print(f'Оптимальные гиперпараметры: {best_params}')

{'n_estimators': 10, 'max_depth': 1, 'min_samples_split': 2, 'min_samples_leaf': 10}
{'n_estimators': 10, 'max_depth': 1, 'min_samples_split': 6, 'min_samples_leaf': 8}
{'n_estimators': 10, 'max_depth': 1, 'min_samples_split': 10, 'min_samples_leaf': 6}
{'n_estimators': 10, 'max_depth': 3, 'min_samples_split': 8, 'min_samples_leaf': 6}
{'n_estimators': 10, 'max_depth': 3, 'min_samples_split': 8, 'min_samples_leaf': 8}
{'n_estimators': 10, 'max_depth': 5, 'min_samples_split': 8, 'min_samples_leaf': 4}
{'n_estimators': 10, 'max_depth': 5, 'min_samples_split': 8, 'min_samples_leaf': 6}
{'n_estimators': 10, 'max_depth': 7, 'min_samples_split': 8, 'min_samples_leaf': 2}
{'n_estimators': 10, 'max_depth': 7, 'min_samples_split': 8, 'min_samples_leaf': 4}
{'n_estimators': 10, 'max_depth': 9, 'min_samples_split': 6, 'min_samples_leaf': 10}
{'n_estimators': 10, 'max_depth': 9, 'min_samples_split': 8, 'min_samples_leaf': 2}
{'n_estimators': 10, 'max_depth': 11, 'min_samples_split': 6, 'min_sample

Модель показала результат на уровне других подходов, не вызвавших подозрение, будет интересно посмотреть на результаты тестовой выборки.

### Выводы по борьбе с дисбалансом
Техники борьбы с дисбалансом привели к увеличению метрик f1 и AUC-ROC для всех моделей. Однако для Дерева решений и Случайного леса количество правильных предсказаний единицы практически стало равным 1, что очень подозрительно. Возможно модели переобучились за счет дублирования данных. В связи с этим на данный момент невозможно выбрать лучшую модель. Использование продвинутого увеличения выборки не привело к максимальному результату, возможно инструмент не совсем правильно используется. Увеличение выборки без кросс-валидации и проверка метрики на валидационной выборке отработала на уровне других подходов. Проверим как все обученные модели будут себя вести на тестовой выборке.

## Тестирование модели

### Дерево решений

In [83]:
print(f'Дисбаланс. f1 = {f1_score(decision_tree_model_disbalance.predict(features_test), target_test)}')
print(f'Взвешивание. f1 = {f1_score(decision_tree_model_balance.predict(features_test), target_test)}')
print(f'Увеличение. f1 = {f1_score(decision_tree_model_upsample.predict(features_test), target_test)}')
print(f'imblearn SMOTE. f1 = {f1_score(decision_tree_model_imba.predict(features_test), target_test)}')
print(f'Уменьшение. f1 = {f1_score(decision_tree_model_downsample.predict(features_test), target_test)}')

Дисбаланс. f1 = 0.5507853403141363
Взвешивание. f1 = 0.586053412462908
Увеличение. f1 = 0.5263157894736842
imblearn SMOTE. f1 = 0.5991416309012875
Уменьшение. f1 = 0.5933384972889234


Все модели превзошли базовую. На тестовой выборке лучше всего себя показала модель, обученная с применением техники продвинутого увеличения выборки, значение метрики f1 достигло 0.596. Также оправдались опасение по поводу увеличения и уменьшения выборки, на тестовой выборке модели показала себя плохо. Дисбаланс, взвешивание и продвинутое увеличение показали значения метрик близкие к полученным на обучающей выборке.

### Случайный лес

In [84]:
print(f'Дисбаланс. f1 = {f1_score(random_forest_model_disbalance.predict(features_test), target_test)}')
print(f'Взвешивание. f1 = {f1_score(random_forest_model_balance.predict(features_test), target_test)}')
print(f'Увеличение. f1 = {f1_score(random_forest_model_upsample.predict(features_test), target_test)}')
print(f'imblearn SMOTE. f1 = {f1_score(random_forest_model_imba.predict(features_test), target_test)}')
print(f'Уменьшение. f1 = {f1_score(random_forest_model_downsample.predict(features_test), target_test)}')
print(f'Увеличение без CV. f1 = {f1_score(random_forest_model_no_cv.predict(features_test), target_test)}')

Дисбаланс. f1 = 0.5926829268292684
Взвешивание. f1 = 0.6292585170340681
Увеличение. f1 = 0.6153846153846153
imblearn SMOTE. f1 = 0.609375
Уменьшение. f1 = 0.6031746031746031
Увеличение без CV. f1 = 0.626552053486151


Все модели превзошли базовую. Для случайного леса взвешивание классов и увеличение выборки без кросс-валидации дали рпактически одинаковые значения, равные примерно 0.62. Увеличение и уменьшение выборки опять не оправдало себя, превзойдя лишь исходные выборки. Дисбаланс, взвешивание, продвинутое увеличение и увеличение без кросс-валидации  показали значения метрик близкие к полученным на обучающей выборке.

### Логистическая регрессия

In [85]:
print(f'Дисбаланс. f1 = {f1_score(logistic_regression_model_disbalance.predict(features_test), target_test)}')
print(f'Взвешивание. f1 = {f1_score(logistic_regression_model_balance.predict(features_test), target_test)}')
print(f'Увеличение. f1 = {f1_score(logistic_regression_model_upsample.predict(features_test), target_test)}')
print(f'imblearn SMOTE. f1 = {f1_score(logistic_regression_model_imba.predict(features_test), target_test)}')
print(f'Уменьшение. f1 = {f1_score(logistic_regression_model_downsample.predict(features_test), target_test)}')

Дисбаланс. f1 = 0.3202328966521106
Взвешивание. f1 = 0.5013736263736264
Увеличение. f1 = 0.5027173913043478
imblearn SMOTE. f1 = 0.494018296973962
Уменьшение. f1 = 0.5026954177897573


Для логистической регрессии все техники борьбы с дисбалансом показали себя примерно одинаково, значение метрики f1 во всех случаях примерно равно 0.5. Без борьбы с дисбалансом логистическая регрессия показала себя хуже базовой. Дисбаланс, взвешивание и продвинутое увеличение показали значения метрик близкие к полученным на обучающей выборке, а увеличение и уменьшение выборки значительно уменьшили показания.

Для наилучшей модели также посчитаем recall.

In [86]:
recall = recall_score(random_forest_model_balance.predict(features_test), target_test)
print(f'Recall = {recall}')

Recall = 0.6421267893660532


### Вывод
Слишком оптимистичные значения метрики f1 при увеличении и уменьшении выборок не зря показались подозрительными. На тестовой выборке модели, обученные с применением данных техники показали плохие результаты. В итоге наилучшее значение 0.62 получила модель случайного леса, обученная с помощью встроенного в sklearn взвешивания классов. Recall этой модели также составил 0.62. 

Применение самодельных техник увеличения и уменьшения выборок не оправдало себя. Также удалост убедиться, что техника увеличения выборки не работает с кросс-валидацией, модель переобучается и показывает плохие результаты на тестовой выборке. В то же время при использовании валидационной выборки мы не даем переобучиться модели и на тестовой выборке она выдает схожие результаты.

## Выводы

Были исследованы данные по оттоку клиентов банка, произведена подготовка данных для обучения моделей и обучены модели с целью максимизации метрики f1.

В данных обнаружен сильный дисбаланс, ушедших клиентов было в 4 раза меньше чем оставшихся. В связи с этим были применены методы борьбы с дисбалансом - взвешивание классов, увеличение и уменьшение выборки, продвинутое увеличение выборки. 

При увеличении и уменьшении выборок на этих выборках модели стали показывать слишком хороший результат, что вызвало подозрение. При их проверке на тестовой выборке подозрения оправдались, показатели моделей упали. В то же время взвешивание и продвинутое увеличение показали значения метрик близкие к полученным на обучающей выборке.


В итоге на тетовой выборке удалось достичь значения метрики f1 равного 0.62 с помощью подбора гиперепараметров для модели случайного леса, дисбаланс был устранен с помощью взвешивания классов. Оптимальными оказались следующие гиперпараметры - {'n_estimators': 190, 'max_depth': 16, 'min_samples_split': 10, 'min_samples_leaf': 4}. Recall модели составил 0.62, соответственно при внедрении модель позволит выявить 62% потенциально уходящих клиентов.


Вместе с измерением f1 измерялась метрика AUC-ROC, было замечено, что с ростом f1 росло и значение AUC-ROC.

## Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  Jupyter Notebook открыт
- [x]  Весь код выполняется без ошибок
- [x]  Ячейки с кодом расположены в порядке исполнения
- [x]  Выполнен шаг 1: данные подготовлены
- [x]  Выполнен шаг 2: задача исследована
    - [x]  Исследован баланс классов
    - [x]  Изучены модели без учёта дисбаланса
    - [x]  Написаны выводы по результатам исследования
- [x]  Выполнен шаг 3: учтён дисбаланс
    - [x]  Применено несколько способов борьбы с дисбалансом
    - [x]  Написаны выводы по результатам исследования
- [x]  Выполнен шаг 4: проведено тестирование
- [x]  Удалось достичь *F1*-меры не менее 0.59
- [x]  Исследована метрика *AUC-ROC*